mempool/src/lib.rs

```diff
@@ -54,14 +54,14 @@ pub fn make_mempool<M>(
     chainstate_handle: subsystem::Handle<Box<dyn ChainstateInterface>>,
     time_getter: TimeGetter,
     memory_usage_estimator: M,
-) -> crate::Result<Box<dyn MempoolInterface>>
+) -> MempoolInterfaceImpl<M>
```

Comment:
While this is better from a pure programmatic point of view, where we have less Box usage (even though compilers can optimize them nowadays), I actually prefer returning `Box<dyn MempoolInterface>` because that enforces adhering to the interface and not using some functionality the interface doesn't provide. As you probably have noticed, I've done this in all subsystems, so that we never have this problem in the future. A middle-ground solution is that you return an `impl MempoolInterface` on the stack instead of a Box, this way both problems are solved. No Box + no bypassing the interface methods.

Reply:
Understood. However in this case, however, `impl MempoolInterface` is not sufficient, since the method to start the subsystem task needs to be exposed too to start the subsystem.

In the node setup code (modified to emphasise the relevant parts):
```rust
let mempool_obj: MempoolInterfaceImpl<_> = make_mempool(/* args */);
let mempool: subsystem::Handle<dyn MempoolInterface> =
    manager.add_raw_subsystem("mempool", move |c, s| mempool_obj.run(c, s));
```

A couple of points here:
* The `run` method needs to be temporarily exposed to register the subsystem, and I don't think it's appropriate to add it to `MempoolInterface`.
* I added type annotations to make it clear that the mempool subsystem handle type is `subsystem::Handle<dyn MempoolInterface>`, so the scope where the concrete type is exposed is very limited. Once you get the subsystem handle, only `dyn MempoolInterface` is exposed.

I can try to hide `MempoolInterfaceImpl` a bit more, maybe by adding a trait that exposes the `run` method and has `MempoolInterface` as a supertrait. All that while only exposing `MempoolInterface` from the handle. I am not sure if it would work but can give this a shot. If it does not, I still think it's not a huge deal since the amount of exposure `MempoolInterfaceImpl` gets is so limited.

Reply:
I now added an extra trait so that the `MempoolInterfaceImpl` is not exposed